File: go/hashmap/383_can_construct.go
package hashmap

// canConstruct 赎金信
func canConstruct(ransomNote string, magazine string) bool {
	// HashMap
	if len(magazine) == 0 || len(magazine) < len(ransomNote) {
		return false
	}

	hmap := make(map[rune]int)

	for _, v := range magazine {
		hmap[v]++
	}

	for _, v := range ransomNote {
		// if cnt, ok := hmap[v]; ok {
		// 	if cnt <= 0 {
		// 		return false
		// 	} else {
		// 		hmap[v]--
		// 	}
		// } else {
		// 	return false
		// }

		hmap[v]--
		if hmap[v] < 0 {
			return false
		}
	}
	return true
}

File: rust/src/hashmap/_383_can_construct.rs


/// 赎金信
pub fn can_construct(ransom_note: String, magazine: String) -> bool {
    use std::collections::HashMap;

    if magazine.len() == 0 || magazine.len() < ransom_note.len() {
        return false;
    }

    let mut hmap = HashMap::new();

    for ch in magazine.chars() {
        hmap.entry(ch).and_modify(|v| *v += 1).or_insert(1);
    }

    for ch in ransom_note.chars() {
        if let Some(cnt) = hmap.get_mut(&ch) {
            *cnt -= 1;
            if *cnt < 0 {
                return false;
            }
        } else {
            return false;
        }
    }

    true
}
